ipa-core/src/query/processor.rs

```diff
@@ -349,17 +361,60 @@ impl Processor {
         Some(status)
     }
 
-    /// Returns the query status.
+    /// Returns the query status in this helper, by querying all shards.
     ///
     /// ## Errors
     /// If query is not registered on this helper.
     ///
     /// ## Panics
     /// If the query collection mutex is poisoned.
-    pub fn query_status(&self, query_id: QueryId) -> Result<QueryStatus, QueryStatusError> {
+    pub async fn query_status(
+        &self,
+        shard_transport: ShardTransportImpl,
+        query_id: QueryId,
+    ) -> Result<QueryStatus, QueryStatusError> {
+        let shard_index = shard_transport.identity();
+        if shard_index != ShardIndex::FIRST {
+            return Err(QueryStatusError::NotLeader(shard_index));
+        }
+
         let status = self
             .get_status(query_id)
             .ok_or(QueryStatusError::NoSuchQuery(query_id))?;
+
+        let shard_query_status_req = CompareStatusRequest { query_id, status };
+
+        shard_transport.broadcast(shard_query_status_req).await?;
```

Comment:
Using errors to communicate between shards is a workaround of our transport trait limitation but we don't have to do it with report collector. The error they will be getting is confusing - they asked for a query status, but got an error "DifferentStatus". We should report the "min" status back to them - if all shards except one is ready, we tell them pending. If at least one error occurred, we say error, etc.

Reply:
I agree. I wasn't sure about this approach for the RC collection. I only went for it because it was a clear implementation and the alternative didn't seemed a lot better. What I had in mind was a new `QueryStatus` called `MixedStatus`. This one was only used when returning for this global state. It didn't felt al that alright to have a state that was only used to reply to a global (sharded) state, not internally. Let me know if this is the route you want to take.

A more involved solution, depending on the current status of the leader vs the broadcast response would look something like the table below:


| Leader Status  | Response if DifferentStatus | Response If query isn't found | Notes |
| ------------- | ------------- | --- | --- |
| None  |  | special | If query IS found in other shard but not in leader we should return an `InconsistentGlobalStateError` (new) |
| Preparing  | Preparing  | Preparing | maybe the query status got earlier than the prepare_shard |
| Awaiting Inputs | Awaiting Inputs | `InconsistentGlobalStateError` |  |
| Running | Running | `InconsistentGlobalStateError` |  |
| Awaiting Completion | Running | `InconsistentGlobalStateError` |  |
| Completed | Running | `InconsistentGlobalStateError` |  |

For this table I'm assuming the helper is going through the states "normally". 

I'm concerned about overcomplicating this thing though.

Reply:
I think we can just use the existing QueryStatus enum and encode these transitions somewhere. It does not feel that this needs to be complicated

Reply:
I'm not prescribing we're the implementation will be, just trying to clarify what each status will be. Can you please be more specific on the "complications"?

Reply:
> I'm concerned about overcomplicating this thing though. 

it is basically a response to this

Reply:
One option is to not use the fact that a query exists (or not) in other shards. 

What I don't like about that though is that the RC will be able to submit queries even if some shard is still stuck in some state (because the leader won't be checking for query existence).

Any specific suggestions?

Reply:
I think just returning an error in this case may be enough for now? as we said failure recovery is complicated and we may want to revisit it later

Reply:
So, instead of sending an error back, we just wrap the error into a non-error response message back?